README.md
<div align="center" style="text-align:center;">

# KSControl
### WiFi-controlled and autonomous gadget to control the Klarstein NB series A/C

<br>

##### Disclaimer:
##### This was only tested with the NB9 model but it'll probably work with other NB series aircons

</div>

<br><br>

### Table of content:

- [Software installation](#software-installation)
- [Hardware installation](#hardware-installation)
  - [Parts list](#parts-list)
  - [Schematic](#schematic)
- [Server](#server)
  - [Set state](#post-state)
  - [Get state](#get-state)
  - [Errors](#errors)

<br><br><br>

## Software installation:
1. Set up the ESP board and firmware in the Arduino IDE ([follow this guide](https://randomnerdtutorials.com/how-to-install-esp8266-board-arduino-ide/))
2. Clone or download and extract this project and open it with the Arduino IDE
3. Install the following library in the library manager (Sketch > Include library): `ESP_EEPROM` by j-watson
4. Rename the file `settings.h.template` to `settings.h`, enter your WiFi's credentials and edit the other settings to your liking
5. Upload the sketch to the ESP board with the arrow button at the top
6. Open the serial monitor ( <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>M</kbd> ) to view information and a live log

<br><br><br>

## Hardware installation:

### Parts list:
- ESP8266 board (or other ESP)
- Infrared LED (100mA, ~1.5V)
- Generic npn transistor
- 30-45 Ω LED resistor (3x 100Ω in parallel or 4x 10Ω in series)
- 1 kΩ transistor resistor

<br>

### Schematic:
<a href="./resources/Schematic/Schematic.svg"><img alt="schematic image" src="./resources/Schematic/Schematic.svg" height="400" /></a>

<br><br><br>

## Server:
KSControl spins up an HTTP server on your ESP's WiFi.  
It can be used to remote control the A/C from within your network.  
To access it from anywhere, you can either set up a VPN connection to your network, host a reverse proxy like nginx on a Raspberry Pi, or forward the port in your router's configuration (insecure, so not recommended!).  
  
Use the following routes to talk to KSControl, using its IP address displayed in the Arduino IDE serial monitor.  

<br>

> ### POST `/state`
> All control data (temp, mode, etc) needs to be simultaneously transmitted to the A/C whenever only one value changes, so the ESP keeps track of the current state of the A/C.  
> If a request is sent to this route, the internal state gets updated and sent to the A/C unit via infrared, just like the remote control would.  
>   
> Some values depend on your A/C model's capabilities (some can't heat for example) and will not influence the actual A/C unit's state.  
> However the ESP's state will be updated with whatever you throw at it, so do some validation yourself if needed.  
>   
> To update the state with new values, send a POST request to this route with the following URL parameters.  
> For example: `/state?enabled=true&mode=COOL&temp=18`
> 
> | Key | Type | Possible values |
> | :-- | :-- | :-- |
> | enabled | boolean | true, false |
> | temp | number | 16-30 °C |
> | mode | string | AUTO, COOL, DRY, FAN, HEAT |
> | fan | string | AUTO, LOW, MEDIUM, HIGH |
>   
> This route returns a JSON object that informs you of the success:
> ```json
> {
>   "error": false,
>   "message": "Successfully updated the state"
> }
> ```

<br>

> ### GET `/state`
> This route returns the current state that's set in the ESP as a JSON object.  
> When no state update has been received by the ESP yet, it will return an empty object.  
>   
> Sadly the ESP's state and the actual A/C unit's state can differ, as someone can use the remote or press the buttons on the unit to change its state without the ESP knowing of it.  
> But if you send a request to the [POST /state](#post-state) endpoint with all parameters set, you will make the states consistent again.  
>   
> The returned data could look like this:
> ```json
> {
>   "error": false,
>   "state": {
>     "enabled": true,
>     "temp": 20,
>     "mode": "COOL",
>     "fan": "HIGH"
>   }
> }
> ```

<br>

### Errors
Client errors have the HTTP response codes 4xx, server errors have the codes 5xx.  
Additionally, KSControl will return a JSON object looking like this, informing you what the problem is:
```json
{
  "error": true,
  "message": "Something went wrong"
}
```

<br><br><br>

<div align="center" style="text-align: center;">

Made with ❤️ by [Sv443](https://github.com/Sv443)  
If you like this project, please consider [supporting me](https://github.com/sponsors/Sv443)  
  
© 2022 Sv443 - [MIT license](./LICENSE.txt)

</div>


changelog.md
### Table of contents:
- [Roadmap](#roadmap)
- **[v0.1.0](#010)**

<br><br>

### Roadmap:
- Add route to poll optional DHT temp & humid sensor
- Support all features (like fan rotation & sleep mode)
- Add automatic on/off timer
    - Change time with HTTP requests
    - Poll NTP server for current time instead of using RTC module

<br>

### 0.1.0
- Basic control for enabled, mode, temperature and fan via HTTP requests


server.cpp
#include "server.h"

// vars

ESP8266WebServer server(SERVER_PORT);

struct AcState state = {true, 20, "AUTO", "AUTO"};

int modeMapLen = 5;
String modeMap[5] = {"AUTO", "COOL", "DRY", "FAN", "HEAT"};
int fanMapLen = 4;
String fanMap[4] = {"AUTO", "LOW", "MEDIUM", "HIGH"};

// ------------------------- #MARKER server -------------------------

// routes
void handleSetState();

void initServer()
{
  EEPROM.begin(EEPROM_SIZE);

  server.on("/state", handleSetState);

  server.begin();
  Serial.print("Started server on port ");
  Serial.println(SERVER_PORT);
}

void handleClient()
{
  server.handleClient();
}

void respond(int status, String body)
{
  server.send(status, "application/json", body);
}

void respondError(int status, String message)
{
  respond(status, "{\"error\":true,\"message\":\"" + message + "\"}");
}

// ------------------------- #MARKER IR -------------------------

int getIRData(AcState state)
{
  int num = 0x50000000;

  auto addBitFieldVal = [&](int n, int val, int pos){
    return n | (val << pos);
  };

  int mode = 0;
  for(int i = 0; i < modeMapLen; i++)
    if(modeMap[i] == state.mode)
      mode = i;
  num = addBitFieldVal(num, mode, 0);
  // Serial.println("Mode: " + String(mode) + " > " + String(num));

  if(state.enabled)
    num = addBitFieldVal(num, 1, 3);
  // Serial.println("Enabled: " + String(num));

  int fan = 0;
  for(int i = 0; i < fanMapLen; i++)
    if(fanMap[i] == state.fan)
      fan = i;
  num = addBitFieldVal(num, fan, 4);
  // Serial.println("Fan: " + String(fan) + " > " + String(num));

  num = addBitFieldVal(num, state.temp - 16, 8);
  // Serial.println("Temp: " + String(num));

  return num;
}

void sendIR(AcState state)
{
  Serial.print("> Sending state via IR // ")
  Serial.print("enabled=" + (state.enabled ? "true" : "false"));
  Serial.print("; temp=" + String(state.temp));
  Serial.print("; mode=" + state.mode);
  Serial.println("; fan=" + state.fan);

  int irData = getIRData(state);

  Serial.println("IR Data: " + String(irData));
}

// ------------------------- #MARKER EEPROM -------------------------

void setState(AcState state)
{
  EEPROM.put(EADDR_STATE, state);

  if(!EEPROM.commit())
    Serial.println("\n---- Error while updating local state in EEPROM ----\n");
}

AcState getState()
{
  struct AcState st;
  EEPROM.get(EADDR_STATE, st);
  return st;
}

// ------------------------- #MARKER routes -------------------------

void handleSetState()
{
  if(server.args() < 1)
    return respondError(400, "Please provide URL parameters");

  auto hasArg = [&](String name)
  {
    return server.hasArg(name);
  };

  auto isBetween = [&](int val, int min, int max){
    return val >= min && val <= max;
  };

  auto includes = [&](String name, String val){
    if(name == "mode")
      for(int i = 0; i < modeMapLen; i++)
        if(modeMap[i] == val)
          return true;
    else if(name == "fan")
      for(int i = 0; i < fanMapLen; i++)
        if(fanMap[i] == val)
          return true;
    return false;
  };

  auto argValid = [&](String name)
  {
    if(!hasArg(name))
      return false;

    String val = server.arg(name);

    if(name == "enabled" && (val == "true" || val == "false"))
      return true;
    else if(name == "temp" && isBetween(val.toInt(), 16, 30))
      return true;
    else if(name == "mode" && includes("mode", val))
      return true;
    else if(name == "fan" && includes("fan", val))
      return true;

    return false;
  };

  auto argInvalid = [&](String name)
  {
    return respondError(400, "Argument \\\"" + name + "\\\" is invalid. Please refer to the documentation.");
  };

  if(hasArg("enabled"))
  {
    if(argValid("enabled"))
      state.enabled = server.arg("enabled") == "true";
    else
      return argInvalid("enabled");
  }
  if(hasArg("temp"))
  {
    if(argValid("temp"))
      state.temp = server.arg("temp").toInt();
    else
      return argInvalid("temp");
  }
  if(hasArg("mode"))
  {
    if(argValid("mode"))
      state.mode = server.arg("mode");
    else
      return argInvalid("mode");
  }
  if(hasArg("fan"))
  {
    if(argValid("fan"))
      state.fan = server.arg("fan");
    else
      return argInvalid("fan");
  }

  setState(state);
  sendIR(state);

  respond(200, "{\"message\":\"Success\"}");
}
